notes/fomega/evaluation/builtin-evaluation.md

```diff
@@ -0,0 +1,347 @@
+   or supplied as an argument to another function.  This doesn't
+   fit in with the way that stack frames are used.
+
+   In more detail, suppose we were to add add a frame `(builtin bn
+   count args)` which accumulates arguments and decrements the
+   counter every time a new argument is supplied.  If the application
+   is saturated then when we get the final argument we can apply the builtin,
+   using a rule along the lines of 
+   
+   `(builtin b 1 args), s ◅ V   ↦   s ◅ W`
+
+   where `b` evaluates to `W` on `reverse (V::args)`.
+
+   But what happens if the application is unsaturated? We might have
+   something like
+
+   `s ▻ [(λf:int->int . f (con 22)) [(builtin addInteger) (con 11)]]`
+
+   You would want this to bind the partially applied builtin to f and then
+   apply it to `(con 22)`, returning 33.  However, it would in fact evaluate
+   as follows:
+   
+   ```
+   s ▻ [(λf:int->int . f (con 1)) [(builtin addInteger) (con 11)]]
+   s, [_, [(builtin addInteger) (con 11)]]  ▻ (λf:int->int . f (con 1))
+   s, [_, [(builtin addInteger) (con 11)]]  ◅ (λf:int->int . f (con 1))
+   s, [(λf:int->int . f (con 1)), _] ▻ [(builtin addInteger) (con 11)]
+   s, [(λf:int->int . f (con 1)), _], (builtin addInteger 2 []) ▻ (con 11)
+   s, [(λf:int->int . f (con 1)), _], (builtin addInteger 2 []) ◅ (con 11)
+   s, [(λf:int->int . f (con 1)), _], (builtin addInteger 1 [(con 11]) ▻ ???
+   ```   
+
+   At this point we're stuck: the `builtin` frame expects a new
+   argument, but nothing is available.  In the version involving
+   saturated builtins, the `builtin` frame contains a list of as-yet
+   unevaluated arguments, and in a situation like this the next one
+   would be extracted and computed at `???`.  We could try to mimic
+   this with unsaturated builtins, but when we run out of immediately
+   available arguments (ie, ones which were supplied to the builtin at
+   invocation time) we would have to convert the frame into some kind
+   of value `W` and compute `s, [(λf:int->int . f (con 1)), _] ▻ W`.
+   This would work, but is pointless: once we've introduced a new type
+   of value for builtins, we no longer need the special `builtin` frame:
+   we can just store the values in the existing frames.  This leads
+   us on to proposal 2.
+      
+
+ 2. We could modify the notion of value used in the CK machine. The CK
+   machine and the CEK machine both have two main phases: "compute"
+   and "return", the former acting on terms and the latter acting on
+   values.  In the CK machine, we've been using values of the type
+   described in the Plutus Core specification, which are terms with a
+   particular syntactic form.  However, there's no particular reason why
+   the type of the arguments consumed by the return phase has to bear any
+   relation to the term type consumed by the compute phase, so this
+   might be the way to go.
+
+   UPDATE: I've now done this and it seems to perform OK: it may be 1%
```

Comment:
I think I am arguing for replacing "Value" with "CKValue".

Reply:
> I think the conversation is evolving in several different places so apologies for chiming in with comments on already covered stuff.

That's fine: it's definitely quite difficult to follow what's going on though with this non-linear conversation and further comments in the PRs for the machines.